### ISSUE TITLE
Doom's performance optimizations break `org-drill` (in general: certain usages of `org-map-entries`)

### ISSUE THREAD

**Issue by RBckmnn** [doomemacs/doomemacs#7979]
### I confirm that...

- [X] I have searched the [issue tracker](https://github.com/doomemacs/doomemacs/issues), [documentation](https://docs.doomemacs.org), [FAQ](https://docs.doomemacs.org/-/faq), [Discourse](https://discourse.doomemacs.org), and [Google](https://google.com), in case this issue has already been reported/resolved.
- [X] I have read "[How to Debug Issues](https://doomemacs.org/d/how2debug)", and will use it to provide as much information about this issue as possible.
- [X] The issue can be reproduced on the **latest** available commit of Doom.
- [X] The issue can be reproduced on a stable release of Emacs, such as 27, 28, or 29. *(Unstable versions end in .50, .60, or .9x)*

### Expected behavior

I ran into this issue while using the `org-drill` package (but it seems to be a general problem). The package allows for spaced repetition practice of flash cards. A flash card corresponds to a heading in an org file. During a drill session a part of the content of such a heading is shown to the user and the user then has to indicate how well he remembered the part not displayed.

### Current behavior

I have noticed that `org-drill` behaved weirdly for quite some time now. The most common problems are that in a practice session `org-drill` sometimes shows headings in a collapsed manner (the contents of the heading should be expanded) or it shows parts of the org file that do not correspond to a heading at all.  Please see below for my attempts to resolve the issue.

### Steps to reproduce

I have produced a minimal working example that exposes the underlying issue. To reproduce:

1. Download the added file, rename it to `file1.org`
2. Define the following function:
```
(defun foo ()
  (interactive)
  (org-map-entries
   (lambda ()
     (sit-for 0.5)
     (org-id-get-create))
   nil '("./file1.org")))

```
3. Open the file downloaded in 1. Make sure that both headlines in the file are collapsed and point is positioned at the beginning of the file. Then invoke the `foo` function.

The function should add an id property to every heading in the file but it does so only for the second heading in the file.
The root cause of this seems to be the `+org--restart-mode-h` function which runs because `org-map-entries` calls `org-agenda-prepare-buffers` in its implementations which kicks off Doom's performance optimizations for org. In fact, turning `+org--restart-mode-h` into a no-op seems to solve all the issues with `org-drill`.
[file1.txt](https://github.com/user-attachments/files/16482966/file1.txt)

### System Information

https://pastebin.com/VavSwE7T

**Comment by hlissner** [comment]
Does this fix the issue for you?

```elisp
;;; add to $DOOMDIR/config.el
(defadvice! fixed-+org--optimize-backgrounded-agenda-buffers-a (fn file)
  :around #'+org--optimize-backgrounded-agenda-buffers-a
  (if-let (buf (org-find-base-buffer-visiting file))
      buf
    (funcall fn file)))
```

I realize the optimization isn't really needed for org buffers that are already open. If this advice works, It'll inform a proper solution for the module.

**Comment by RBckmnn** [comment]
Hello Henrik,

thanks for you speedy reply. The advice you proposed caused an error. I assume you meant

```
;; add to $DOOMDIR/config.el
(defadvice! fixed-+org--optimize-backgrounded-agenda-buffers-a (fn1 fn2 file)
  :around #'+org--optimize-backgrounded-agenda-buffers-a
  (if-let (buf (org-find-base-buffer-visiting file))
      buf
    (funcall fn1 fn2 file)))
```

It seems that this fixes my minimal example. However, `org-drill` still seems to be broken, probably because it supports drilling from any number of org files (which are not necessarily opened yet).

(see the definition of `org-drill-scope` [here](https://gitlab.com/phillord/org-drill/-/blob/master/org-drill.el?ref_type=heads#L342) and its usage in `org-drill-map-entries` [here](https://gitlab.com/phillord/org-drill/-/blob/master/org-drill.el?ref_type=heads#L708) )

**Comment by hlissner** [comment]
This actually doesn't related to `+org--optimize-backgrounded-agenda-buffers-a`, but more on where the cursor is in already-open agenda buffers. I.e. If the cursor is on the second heading, `org-map-entries` only operates on that heading (and any after it), but if the cursor is at the beginning of the buffer, it operates on all headings correctly. This occurs whether or not the `+org--optimize-backgrounded-agenda-buffers-a` advice exists.


I suspect `saveplace`, but I'm looking more into it.

**Comment by hlissner** [comment]
Oh, this might actually be `sit-for`'s fault. If you swap it out for `sleep-for`, `foo` works in all cases. I suspect because it triggers a redisplay which has side-effects that could trigger all sorts of hooks.

Do you see this too?

```diff
 (defun foo ()
   (interactive)
   (require 'org)
   (org-map-entries
    (lambda ()
-     (sit-for 0.5)
+     (sleep-for 0.5)
      (org-id-get-create))
    nil '("./file1.org")))
```

**Comment by RBckmnn** [comment]
Yes, the call to `sit-for` was essential for my minimal example to exhibit the behavior (apologies, I could have been clearer in that regard). Unfortunately, such calls to `sit-for` are quite common in the `org-drill` package and even if one fixes the issue with the `org-map-entries` callback, the package is still not working correctly. I have extracted another example from the `org-drill` source code:


``` emacs-lisp
(defun foo/minimal-example ()
  (interactive)
  (require 'org)
  (let ((markers (org-map-entries
                  (lambda () (point-marker))
                  nil '("./file1.org"))))
    (let ((m (car markers)))
      (switch-to-buffer (marker-buffer m))
      (goto-char m)
      (org-fold-show-entry)) ;; Heading is expanded here
    (foo/wait-for-keysequence) ;; Inside this call the heading will be collapsed again
    (dolist (m markers) (set-marker m nil))))


(defun foo/wait-for-keysequence ()
  (let ((item-start-time (current-time))
        (input nil)
        (ch nil))
    (while (null ch)
      (setq ch nil)
      (while (not (input-pending-p))
        (let ((elapsed (time-subtract (current-time) item-start-time)))
          (message "Elapsed %s. Waiting for input" (format-time-string "%M:%S" elapsed))
          (sit-for 1)))
      (setq input (read-key-sequence nil))
      (if (stringp input) (setq ch (elt input 0))))))
```

Assume that `foo/minimal-example` is called and the `file1.org` file is not yet opened. The call to `org-fold-show-entry` expands the heading at point. However, the subsequent call of `sit-for` allows Doom's optimization to run, which as a side effect collapses the heading again. The question seems to be if this should be considered a bug in Doom or in `org-drill`.

**Comment by hlissner** [comment]
Are you still seeing this issue? I suspect that https://github.com/doomemacs/doomemacs/commit/ded3f5ec834942044b5caae6ca220270660b466b might've indirectly fixed it; it caused a lot of strange issues to do with buffer focus and lifecycles that *could* explain the behavior here.

**Comment by RBckmnn** [comment]
Unfortunately, I'm still seeing this issue, even after updating to the latest commit of Doom (94a291a7f8f3425a33a8326077cfb0c9964f1631) . Using the example [above](https://github.com/doomemacs/doomemacs/issues/7979#issuecomment-2282857596) I can still produce issues. However, I realize that the following snippet from my `config.el` might have an impact here:

```elisp
  (after! org
    (setq org-startup-folded 'content))
```

With the following steps you might be able to reproduce the issue:

1. Add the snippet above to your `config.el`
2. Restart emacs
3. Evaluate the elisp code in the example linked above
4. Use `M-x` to invoke `foo/minimal-example`

Expected behavior:
Emacs opens the `file1.org` file. Point is positioned at buffer position 1. The first heading is expanded. The second heading is folded.

Actual behavior:
Emacs open the `file1.org` file. Point is almost never  positioned at position 1. It is sometimes positioned on the beginning  of the second heading. Sometimes point is positioned at a position different from 1 in the first line.  The second heading is sometimes expanded, sometimes not. The first heading is never expanded.

In addition on may perform the following steps: 
5. After invoking `foo/minimal-example`, do any or none of the following in the `file1.org`-buffer : a) Move point, b) change the expansion state of headings in the buffer
6. Kill the buffer
7. Invoke `foo/minimal-example` again

Expected behavior:
As the expected behavior above

Actual behavior:
The position of point and the expansion state of the headings is different from the expected behavior but sometimes also different from the results observed when `foo/minimal-example` was called the first time.

**Comment by hlissner** [comment]
Does wrapping the body of `foo/minimal-example` in:

```elisp
(let (save-place-alist)
  ...)
```

Fix the issue for you?

**Comment by RBckmnn** [comment]
Your first suggestion brought some improvement. Every call to `foo/minimal-example` now has the following deterministic results:

- Point is always positioned at buffer position 1 (expected)
- The first heading is always collapsed (not expected)
- The second heading is always collapsed (expected)

Your second suggestion did not yield any discernible improvements. The behavior of `foo/minimal-example` with regards to position of point and the expansion state of the headings is still erratic.

Thanks a lot for looking into this and apologies for being a little bit of a hassle.
